messaging.rst

```diff
@@ -22,31 +23,56 @@ The further messages within the transfer are based on it:
 
 Further messages in the protocol are:
 
-- The :ref:`Processed <processed-delivered-message>` that is sent to confirm received messages, and
+- The :ref:`Processed and Delivered <processed-delivered-message>` messages to acknowledge received messages, and
 - The withdraw-related messages :ref:`WithdrawRequest <withdraw-request-message>`,
   :ref:`WithdrawConfirmation <withdraw-confirmation-message>` and
   :ref:`WithdrawExpired <withdraw-expired-message>`.
 
-Encoding and transport
-======================
+Encoding, signing and transport
+===============================
 
 All messages are encoded in a JSON format and sent via our Matrix transport layer.
 
 The encoding used by the transport layer is independent of this specification, as
 long as the signatures using the data are encoded in the EVM big endian format.
 
+.. _message-classes:
+
+The messages of the protocol can be divided in three groups with different format/hashing/signing
+conventions:
+
+- **Envelope messages**, which contain a balance proof which can be sent to a contract. The
+  balance proof in turn contains an :term:`additional hash`, which is a hash over the rest of
+  the message. Each envelope message has a defined packed data format to compute the additional
+  hash. The format always starts with the 1-byte command id. Envelope messages are:
+  ``LockedTransfer``, ``RefundTransfer``, ``Unlock`` and ``LockExpired``.
+
+- The second group is messages that will never result in on-chain transactions, as they contain
+  no information that could be forwarded to a contract. There are four types of such messages,
+  which we will call **internal messages**: ``SecretRequest``, ``RevealSecret``, ``Processed``
```

Comment:
You could object to the term "internal messages" by saying that they are transferred between different nodes and are thus not internal. But I don't have a better name and it is unambiguous enough when used in this context, so I don't object.